src/rebar3_hex_publish.erl

```diff
@@ -69,21 +66,45 @@ do(State) ->
             ?PRV_ERROR(Reason)
     end.
 
+%% Revert cases 
 handle_task(#{args := #{revert := undefined}}) ->
     {error, "--revert requires an app version"};
 
-handle_task(#{args := #{revert := Vsn}, repo := Repo, state := State}) ->
+handle_task(#{args := #{revert := Vsn}, multi_project := false} = Task) ->
+    #{repo := Repo, state := State} = Task,
     App = rebar_state:current_app(State),
     Name = rebar_app_info:name(App),
     ok = rebar3_hex_revert:revert(binarify(Name), binarify(Vsn), Repo, State),
     {ok, State};
 
-handle_task(#{repo := Repo, state := State}) ->
+handle_task(#{args := #{revert := Vsn, app := AppName}, multi_project := true} = Task) ->
+    #{repo := Repo, state := State} = Task,
+    ok = rebar3_hex_revert:revert(binarify(AppName), binarify(Vsn), Repo, State),
+    {ok, State};
+
+handle_task(#{args := #{revert := _}, multi_project := true}) ->
+    {error, "--app required when reverting in a release with multiple projects"};
+
+%% Publish 
+
+handle_task(#{repo := Repo, state := State, multi_project := false}) -> 
+    case maps:get(write_key, Repo, maps:get(api_key, Repo, undefined)) of
+            undefined ->
+                ?PRV_ERROR(no_write_key);
+            _ ->
+                App = rebar_state:current_app(State),
+                publish(App, Repo, State)
+        end;
+
+handle_task(#{repo := Repo, state := State, multi_project := true}) ->
```

Comment:
This could be a single clause function if you create a function `select_apps` which when `multi_app` is false simply returns the current app in a list `[App]` and if `multi_app` is `true` then do the `select_app`, then run the `publish` over the list.

Reply:
I think what I was shooting for was making the behavior obvious and up front, maybe that's a bit of overkill / obsessive passion though. TBH I'm still not sure exactly what that's going to look like, I'd have a better idea after doing some refactoring (today). If the results of refactoring show that it's better to have a single function clause, we can go with that, does that sound good?